### PR TITLE
Refactor mutators to typefy inputs | Closes #2399

### LINKED ISSUE #2399: Enable type support for mutation inputs
Currently, `MutationInput` is of this shape:
```ts
type MutationInput = {
	[key: string]: any;
};
```
which makes it vulnerable for mistakes. Making use if TS Generics, we should be able to correctly set its shape.
The next task would be to actually create types for different mutation inputs.

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR does a complete overhaul of how mutators work. Previously, we used `useEntityMutator` from `/application` which was too generic to be locked down for types. One more issue was that it was tightly coupled with domain mutations. This PR solves these issues:
- Replaces `useEntityMutator` by `useMutator` and `useEntityMutation` by `useMutationHandler`
- Creates a `useMutator` hook in `eventEditor` domain to carry out mutations and set proper types
- Creates types for datetime, price and ticket mutations.
- Creates mutators for specific entities viz. `useDatetimeMutator`, `useTicketMutator` & `usePriceMutator` which should be used from now on to carry out  all the mutations.
- TPC submit errors are not fixed yet.
- Closes #2399

### CODE REVIEW

File: assets/src/application/services/apollo/mutations/types.ts [hunk 1]
```diff
@@ -1,80 +1,72 @@
...
+export type MutationOptionsCb = <Type extends BaseType, MI = MutationInput>(
+	mutationType: MutationType,
+	input: MI
+) => MutationOptions<Type>;
+
+export type Mutator = (getMutationOptions: MutationOptionsCb) => MutatorCallback;
+export type MutatorCallback = <Name extends string, Type extends BaseType>(
+	typeName: Name,
+	id?: string
+) => MutatorObject;
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/src/application/services/apollo/mutations/types.ts]: That's just to to match the signature of the actual function that needs it to use internally.

File: assets/src/application/services/apollo/mutations/useMutationHandler.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ * @param id
+ */
+const useMutationHandler: MutationHandler = (getMutationOptions) => {
+	const client = useApolloClient();
+	const [result, setResult] = useState(DEFAULT_RESULT);
+	const ifMounted = useIfMounted();
+	const toaster = useToaster();
+
+	return useCallback(
+		<Name extends any>(typeName: Name, id?: string) => {
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/src/application/services/apollo/mutations/useMutationHandler.ts]: It shouldn't be 👍

File: assets/src/application/services/apollo/mutations/useMutator.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * @param {string} type Entity type name
+ * @param {string} id   Entity id
+ */
+const useMutator: Mutator = (getMutationOptions) => {
+	const handleMutation = useMutationHandler(getMutationOptions);
+
+	return useCallback<MutatorCallback>(
+		<Name extends string, Type extends BaseType>(typeName: Name, id = '') => {
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/src/application/services/apollo/mutations/useMutator.ts]: Yes

File: assets/src/application/services/apollo/mutations/types.ts [hunk 4]
```diff
@@ -1,80 +1,72 @@
...
+}
 
-export type OnUpdateFn = (options: OnUpdateFnOptions) => void;
+export type OnMutationCompletedFn<DShape = any> = (data: DShape) => void;
 
-export interface BackwardSubscription {
-	onCompleted?: OnMutationCompletedFn;
+export type OnMutationErrorFn = (error: Error) => void;
+
+export interface BackwardSubscription<DShape = any> {
```
Comment (nicolad) [inline assets/src/application/services/apollo/mutations/types.ts]: A comment might be helpful here.

File: assets/src/application/services/apollo/mutations/useMutationHandler.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+			/**
+			 * @param {string} mutationType Type of mutation - CREATE|UPDATE|DELETE
+			 * @param {string} suffix       Suffix to be added to mutation typeName
+			 */
+			const getToasterMessage = (mutationType: MutationType, suffix: 'ing' | 'ed' = 'ing'): any => {
+				// For example "CREATE" will become "creating" or "created"
+				const verb = mutationType.toLowerCase().replace(/e$/, suffix);
+				// e.g. "updating datetime"
+				return `${verb} ${typeName.toLowerCase()}`;
```
Comment (nicolad) [inline assets/src/application/services/apollo/mutations/useMutationHandler.ts]: :+1:

File: assets/src/application/services/apollo/mutations/useMutationHandler.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+				updateResult({
+					loading: true,
+					error: undefined,
+					data: undefined,
+					called: true,
+				});
+			};
+
+			/**
+			 *
```
Comment (nicolad) [inline assets/src/application/services/apollo/mutations/useMutationHandler.ts]: Placeholder for future comments?

File: assets/src/application/services/apollo/mutations/types.ts (hunk 4 above)
Comment (manzoorwanijk) [reply assets/src/application/services/apollo/mutations/types.ts]: Done

File: assets/src/application/ui/forms/formModal/types.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { FormProps, FormRenderProps } from 'react-final-form';
 import { FormApi } from 'final-form';
 
 export interface FormModalProps extends FormProps {
-	FormComponent: React.FC;
+	FormComponent: React.NamedExoticComponent;
```
Comment (nicolad) [inline assets/src/application/ui/forms/formModal/types.ts]: Didn't knew about this one :smiley:

File: assets/src/application/services/apollo/mutations/types.ts [hunk 8]
```diff
@@ -53,6 +55,8 @@ export type MutatorFnGn<MI = MutationInput, DShape = any> = (
 	subscriptions?: BackwardSubscription<DShape>
 ) => MutationResult<DShape>;
 
+// These are forward subscriptions towards mutators
+// i.e. when an entity mutation handler wants to be notified
```
Comment (nicolad) [inline assets/src/application/services/apollo/mutations/types.ts]: :+1:

**Pull request merged**